### PR TITLE
Week 5 Development

### PR DESCRIPTION
**Pull request by bradleypickard**
Sorry for the weird cadence this week, but here's my week 5 stuff.

Implements and tests calendar scheduler. Scheduler finds the times during the day when a meeting with given Attendees and Duration could occur, given the Attendees' existing schedules.

Also includes basic Optional Attendees functionality, where the scheduler will at first try to include Attendees categorized as Optional, but if necessary schedule based only on the mandatory Attendees' schedules.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +362,68 @@ public void notEnoughRoom() {
...
+    MeetingRequest request = new MeetingRequest(NO_ATTENDEES, DURATION_2_HOUR);
+    request.addOptionalAttendee(PERSON_A);
+    request.addOptionalAttendee(PERSON_B);
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected = Arrays.asList();
+
+    Assert.assertEquals(expected, actual);
+  }
+}
```
Comment (hanpeiz) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:429]: You should add some edge/boundary cases: e.g. no event has been scheduled at all; Attempt to schedule a meeting with duration=0; etc

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,75 @@
...
+            iter.remove();
+    }
+
+    return(freeList);
+  }
+
+  private void trim(ListIterator iter, TimeRange freeBlock, int eventStart, int eventEnd) {
+      if (eventStart <= freeBlock.start() && eventEnd >= freeBlock.end())   // Event fully contains free-block
+          iter.remove();
+      else if (eventStart <= freeBlock.start()) {   // Event overlaps with beginning of free-block
```
Comment (hanpeiz) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: One simple measurement how good your test cases are is called "line coverage". It is essentially the percentage of lines of code your tests executes. We generally try to get 100% line coverage - i.e. every line that contains some kind of logic is executed in at least one of the test cases. Correct me if I'm wrong, but I don't think you have a test case where an event partially overlap a free-block. I think this only happens when existing events partially overlap each other.

**Review by hanpeiz — COMMENTED** [review]
Looks good overall!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,75 @@
...
+      if (freeList.size() == 0)
+          return performQuery(events, request, false);
+      return freeList;
+  }
+  
+  public Collection<TimeRange> performQuery(Collection<Event> events, MeetingRequest request, boolean considerOptionalAttendees) {
+    LinkedList<TimeRange> freeList = new LinkedList<TimeRange>();
+    freeList.add(TimeRange.WHOLE_DAY);
+
+    // Carve freeList down into series of non-conflicting chunks
```
[the remark this answers is not in the document]
Comment (bradleypickard) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:36]: Will do!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,75 @@
...
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      if (request.getAttendees().size() == 0)
+          return performQuery(events, request, true);
```
[the remark this answers is not in the document]
Comment (bradleypickard) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: I did this in response to some weirdness in the provided test cases.

If you look at test 'optionsForNoAttendees,' when there are no attendees we act as though the whole day is available.

However, according to the walkthrough, when there are
' 
no mandatory attendees, just two optional attendees with no gaps in their schedules, query should return that no time is available.
'

So when we have optional attendees and no mandatory attendees, if we can't find a time-frame that satisfies everyone, we don't (as per usual) default to checking availability for just mandatory. In the above case, doing so would give the whole day.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 1 above)
Comment (bradleypickard) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:429]: I'll add some for the next pr.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (bradleypickard) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: I was sure that was covered somewhere, but I guess I need to read better. I'll be sure this is added!

**Review by hanpeiz — APPROVED** [review]
